src/shim/binary_spinning_sem.h

```diff
@@ -61,6 +61,13 @@ class BinarySpinningSem {
      */
     void wait(bool spin = true);
 
+    /*
+     * Atomically check if the semaphore is available (has value one). If
+     * so takes the semaphore (sets it back to zero) and returns 0. Otherwise
+     * returns -1 and sets errno to EAGAIN.
+     */
+    int trywait();
```

Comment:
n.b. I haven't worked it out, but this might violate thread-safety if called in parallel with wait(). I think that's OK because that's not a use case we're really using, but maybe worth noting.

Reply:
Hmm... I'd rather make this safe than add a potential footgun of `trywait` and `wait` not being safe to use together...

The way I look at it is that `_semaphore` is the authority on whether the BinarySpinningSem is available, and `_x` is only a hint. TBH I don't think I understand what `_lock` is doing.

I changed it to try grabbing `_x` first to avoid a thread calling `_wait`, grabbing `_x`, breaking out of its spin loop, only to have a concurrent thread grab the semaphore out from under it via `trywait`. Does this address your concern?

FWIW a spurious break in `wait` is already (and still) possible when two threads call `wait`: thread A can hit its spin max (or not spin at all) and wait on `_semaphore` without having grabbed `_x` first. After the semaphore is released, another thread calling `wait` could grab `_x` but fail to grab the semaphore, which the first waiter grabbed out from under it. I don't see a way to avoid that problem; perhaps another reason to see if we should reduce to just spinning directly on the semaphore with `sem_trywait`. I'll take a closer look at that, but would rather save that for another PR.